### ISSUE TITLE
Improvements to the HttpClient / HttpServer

### ISSUE THREAD

**Issue by slaff** [slaff/Sming#5]
## Performance and resources usage

(The progress can be seen and tested in [that branch](https://github.com/slaff/Sming/tree/feature/better-http) )

The current HttpClient has multiple disadvantages when the http client is used to make more than one request. These disadvantages can be seeing using the following example:

```
HttpClient client();
client.downloadString("https://google.com/1", onRequestComplete);
client.downloadString("https://google.com/2", onRequestComplete);
client.downloadString("https://google.com/3", onRequestComplete);
```

With the current code most probably only the last request will be executed and all others will be discarded silently. Even if these requests are separated with enough time for each one to finish, then the current client will create 3 new tcp connections and will try 3 times to make full SSL handshake. 

This is suboptimal and can be improved with the following known techniques.

 Name | Description | Advantages | Remarks
--------------|-------------|-----------------|----------------------------
Pipelining | Multiple consecutive GET and  HEAD requests can be sent via one TCP connection without waiting for the answer | Only one tcp connection. And if we use SSL then only one handshake will be done
Connection Reuse | Once the HTTP request is over the TCP connection should be kept open for some seconds so that other http requests can be sent over it. | One tcp connection can be reused for multiple Http requests even if those requests cannot be pipelined
SSL Resumption| SSL handshake is very time, cpu and memory intensive task for embedded devices. Once a success handshake is done we can reuse the ssl session for the same host and port and the consecutive SSL handshakes will use SSL resumption. Usually the SSL session ids are kept for 30min to 1 hour. | Only one full handshake to a host and port will be done once every 30 min - 1hour instead of a full SSL handshake. That SSL session id can be used in different TCP connections that lead to the same host:port.

## Code design
The current HttpServer and HttpClient reimplement almost the same code twice. It will be better if we have the following classes that can be reused and maintained better.

* HttpResponse - that is a data class keeping information about one HttpResponse. This class must not know how the response will be sent, parsed, etc.
* HttpRequest - similar to HttpResponse that class should be only data class.
* HttpConnection - should represent a connection to one `host:port`. In this class the parsing of the response data should be done. 
* HttpClient - a class that takes care to pipeline, reuse TCP connection and SSL session ids for all HTTP requests.
* Separate Http Parser that can parse  HTTP response and request data.

If we have the above classes we can reuse the http parser, HttpResponse and HttpRequest in our HttpServer

### Phase 1
- [x] Create classes alongside the existing HttpClient, HttpResponse and HttpRequest and use `Web` as prefix. Implement the needed HTTP techniques (done in 0795ad97e479441bd060288c13f1c769829cc217)
- [x] and test the code on real device

### Phase 2
- [x] Replace the HttpClient code with the one from WebClient. Remove WebClient.
- [x] Make sure that other classes that extend from HttpClient can also work
- [x] Add missing compatibility methods to HttpClient that were present also in the old version.

### Phase 3
- [x] Add convenience methods to support POST requests and FileUpload requests.

### Phase 4

Improvements to the HttpServer

Name | Description | Advantages | Remarks
--------------|-------------|-----------------|----------------------------
Pipelining | The server should able able to process multiple consecutive GET and  HEAD requests sent via one TCP connection | Only one tcp connection. And if we use SSL then only one handshake will be done
Connection Reuse | The server should not close the connection immediately, unless `Connection: close` request headers is given. KeepAlive should be specified as a server settings parameter

- [x] Replace the simple http parser in HttpServer with the http parser used in the HttpClient code.
- [x] Refactor the HttpServer to use WebRequest and WebResponse classes.
- [x] Replace HttpResponse and HttpRequest with WebResponse and WebRequest. Delete WebResponse and WebRequest code.

### Phase 5
- [x]  Server-side SSL support
- [x] (Nice-to-have) Server-side SSL session id cache. Allow the http clients to be able to resume SSL sessions. Have configuration setting that specifies the ssl session id expiration time.
- [ ] Add optional multipart/form processing (server side mainly)

### Phase 6
- [ ] Documentation
- [ ] Migration Guide
